README.md
![Pub](https://img.shields.io/pub/v/dropdown_banner.svg?style=popout)

# dropdown_banner
DropdownBanner manages the creation and animation of banner ui elements
that are useful for displaying warnings and updates to users. It is simple to use and effective at event communication.

![](demonstration.gif)

## Getting Started
Using the dropdown banner is quite intuitive and simple. Below is a quick example of how to use it!

#### In your main.dart when creating the app
```dart
import 'packages:dropdown_banner/dropdown_banner.dart';
...
class MainApp extends StatelessWidget {
  ...
  @override
  Widget build(BuildContext context) {
    final navigatorKey = GlobalKey<NavigatorState>();
    ...
    return MaterialApp(
        ...
        home: DropdownBanner(
          child: Scaffold(...),
          navigatorKey: navigatorKey,
        ),
    );
  }
}
```

#### Generate a banner event
```dart
import 'packages:dropdown_banner/dropdown_banner.dart';
...
class SomeClass {
  ...
  void doSomethingThenFail() {
    DropdownBanner.showBanner(
      text: 'Failed to complete network request',
      color: SKColors.warning_red,
      textStyle: TextStyle(color: Colors.white),
    );
  }
}
```

## Planned improvements
If this package gets any traction, I plan to add support for images, custom widgets, animation duration.. etc. If you have any ideas for making this package more applicable, just open an issue!

## Other info
This package uses another of my packages, [dart_notification_center](https://pub.dev/packages/dart_notification_center), for implementing the observer pattern. This is what makes using the banner so simple! Check it out if you find a need for it.


CHANGELOG.md
## [1.4.0] - 10/2/2019

* Swipe up to dismiss banner

## [1.3.0] - 9/26/2019

* Dynamically sizeable banners
* Improved library structure
* Maintenance and documentation

## [1.2.0] - 8/19/2019

* Change when library fetches height information

## [1.1.1] - 7/25/2019

* Fixes another issue with the navigator state resetting after every hot reload. should be good to go now. Updated the example to represent new package architecture.
* This update has breaking changes. Sorry for the inconvenience, but this should greatly improve the stability and usability of the package.

## [1.1.0] - 7/25/2019

* Fixes the issue where child is rebuilt everytime navigator is used or hot reload occurs

## [1.0.2] - 7/15/2019

* Enabled navigator key override to prevent hot reloading from resetting

## [1.0.1+1] - 7/11/2019

* Upgraded package dependency version


## [1.0.1] - 7/11/2019

* Added shield and shortened description

## [1.0.0] - 7/11/2019

* Example project
* Demonstration GIF
* Flutter formatter
* Project description

## [0.1.1] - 7/11/2019

* Updated SDK version to enable pub parsing

## [0.1.0] - 7/11/2019

* Initial version of the package.
* Basic documentation completed.


test/dropdown_banner_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:dropdown_banner/dropdown_banner.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:dropdown_banner/dropdown_banner.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  final navigatorKey = GlobalKey<NavigatorState>();

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blueGrey,
      ),
      home: DropdownBanner(
        child: HomeWidget(),
        navigatorKey: navigatorKey,
      ),
    );
  }
}

class HomeWidget extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => _HomeWidgetState();
}

class _HomeWidgetState extends State<HomeWidget> {
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: Text('DropdownBanner Example')),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Spacer(flex: 1),
              Text('Tap an option to see the banner in action'),
              Padding(
                padding: EdgeInsets.symmetric(vertical: 16),
                child: GestureDetector(
                  onTapUp: (details) => DropdownBanner.showBanner(
                    text: 'Successfully saved your info!',
                    color: Colors.green,
                    textStyle: TextStyle(
                        color: Colors.white,
                        fontSize: 16,
                        fontWeight: FontWeight.bold),
                  ),
                  child: Container(
                    padding: EdgeInsets.symmetric(horizontal: 36, vertical: 8),
                    decoration: BoxDecoration(
                      color: Colors.green,
                      borderRadius: BorderRadius.circular(5),
                      boxShadow: [
                        BoxShadow(
                            blurRadius: 5, color: Colors.black.withOpacity(0.2))
                      ],
                    ),
                    child: Text(
                      'Test 1',
                      style: TextStyle(
                          color: Colors.white, fontWeight: FontWeight.bold),
                    ),
                  ),
                ),
              ),
              Padding(
                padding: EdgeInsets.symmetric(vertical: 16),
                child: GestureDetector(
                  onTapUp: (details) => DropdownBanner.showBanner(
                    text: 'Failed to complete the request',
                    color: Colors.red,
                    textStyle: TextStyle(
                        color: Colors.white,
                        fontSize: 16,
                        fontWeight: FontWeight.bold),
                  ),
                  child: Container(
                    padding: EdgeInsets.symmetric(horizontal: 36, vertical: 8),
                    decoration: BoxDecoration(
                      color: Colors.red,
                      borderRadius: BorderRadius.circular(5),
                      boxShadow: [
                        BoxShadow(
                            blurRadius: 5, color: Colors.black.withOpacity(0.2))
                      ],
                    ),
                    child: Text(
                      'Test 2',
                      style: TextStyle(
                          color: Colors.white, fontWeight: FontWeight.bold),
                    ),
                  ),
                ),
              ),
              Padding(
                padding: EdgeInsets.symmetric(vertical: 16),
                child: GestureDetector(
                  onTapUp: (details) => DropdownBanner.showBanner(
                    text:
                        'DropdownBanner has been added to your cart. You now have 15 items. Tap to checkout.',
                    color: Colors.black.withOpacity(0.8),
                    textStyle: TextStyle(
                        color: Colors.white,
                        fontSize: 14,
                        fontWeight: FontWeight.normal),
                    tapCallback: () async => showDialog(
                      context: context,
                      builder: (context) => Dialog(
                        shape: RoundedRectangleBorder(
                            side: BorderSide(color: Colors.black),
                            borderRadius: BorderRadius.circular(10)),
                        child: Container(
                          margin: EdgeInsets.all(16),
                          child: Text(
                            'You did a thing! 🎉',
                            textAlign: TextAlign.center,
                          ),
                        ),
                      ),
                    ),
                  ),
                  child: Container(
                    padding: EdgeInsets.symmetric(horizontal: 36, vertical: 8),
                    decoration: BoxDecoration(
                      color: Colors.black,
                      borderRadius: BorderRadius.circular(5),
                      boxShadow: [
                        BoxShadow(
                            blurRadius: 5, color: Colors.black.withOpacity(0.2))
                      ],
                    ),
                    child: Text(
                      'Test 3',
                      style: TextStyle(
                          color: Colors.white, fontWeight: FontWeight.bold),
                    ),
                  ),
                ),
              ),
              Spacer(flex: 2),
            ],
          ),
        ),
      );
}


lib/dropdown_banner.dart
library dropdown_banner;

import 'dart:async';
import 'package:dart_notification_center/dart_notification_center.dart';
import 'package:flutter/material.dart';

/// Channel name to communicate menu updates on
const String _BANNERCHANNEL = 'createDropdownBanner';

typedef void _IntCallback(int id);

/// DropdownBanner manages the creation and animation of banner elements
/// that are useful for displaying warnings and updates to users.
class DropdownBanner extends StatefulWidget {
  /// Builder in which to construct the app content that you are wrapping
  final Widget child;
  final GlobalKey<NavigatorState> navigatorKey;

  DropdownBanner({@required this.child, @required this.navigatorKey}) {
    DartNotificationCenter.registerChannel(channel: _BANNERCHANNEL);
  }

  /// Used to track banners uniquely
  static int _idCounter = 1;

  /// Display a banner with the desired [text] and [textStyle] on a [color] background
  /// for the [duration] specified. If the banner is tapped and [tapCallback] != null,
  /// the callback will be executed and the banner dismissed.
  static showBanner({
    @required String text,
    Duration duration,
    Color color,
    TextStyle textStyle,
    VoidCallback tapCallback,
  }) {
    DartNotificationCenter.post(
      channel: _BANNERCHANNEL,
      options: {
        'id': _idCounter,
        'duration': duration ?? Duration(seconds: 3),
        'text': text,
        'color': color ?? Colors.white,
        'textStyle': textStyle,
        'tapCallback': tapCallback,
      },
    );

    // Increment counter for next use
    ++_idCounter;
  }

  @override
  State<StatefulWidget> createState() => _DropdownBannerState();
}

class _DropdownBannerState extends State<DropdownBanner> {
  List<_BannerInstance> banners = [];

  @override
  void initState() {
    super.initState();

    // Get notified when a Banner is requested to be created
    DartNotificationCenter.subscribe(
      channel: _BANNERCHANNEL,
      observer: this,
      onNotification: createBanner,
    );
  }

  @override
  void dispose() {
    DartNotificationCenter.unsubscribe(observer: this);

    banners = [];

    super.dispose();
  }

  void createBanner(dynamic bannerDetails) {
    // Ensure that the call to create the banner was from within the Library
    assert(
      bannerDetails is Map<String, dynamic> && bannerDetails['id'] is int,
      'Do not post to $_BANNERCHANNEL using DartNotificationCenter, as this is reserved for internal use for DropdownBanner to work properly.',
    );

    // Add banner to banners, and display
    setState(
      () => banners.add(
        _BannerInstance(
          bannerDetails['id'],
          bannerDetails['duration'],
          bannerDetails['text'],
          bannerDetails['color'],
          bannerDetails['textStyle'],
          bannerDetails['tapCallback'],
          onAnimationCompletion,
        ),
      ),
    );
  }

  /// Callback for removing banner from banner queue
  void onAnimationCompletion(int id) =>
      setState(() => banners.removeWhere((b) => b.id == id));

  @override
  Widget build(BuildContext context) => Stack(
        children: <Widget>[
          Align(
            alignment: Alignment.center,
            child: Navigator(
              key: widget.navigatorKey,
              onGenerateRoute: (settings) => MaterialPageRoute(
                settings: settings,
                builder: (_) => widget.child,
              ),
            ),
          ),
          ...banners
        ],
      );
}

/// Container to track various aspects of the appearance and life of a dropdown banner object
class _BannerInstance extends StatefulWidget {
  /// Identify unique state between rebuilds
  final Key key = UniqueKey();

  /// Unique id for referencing
  final int id;

  /// Length of time the banner stays visible for
  final Duration duration;

  /// Text to display in the banner
  final String text;

  /// Color of the banner
  final Color color;

  /// Style for the text displayed
  final TextStyle textStyle;

  /// Action to perform on tap
  final VoidCallback tapAction;

  /// Callback to remove from parent render tree
  final _IntCallback onCompletion;

  _BannerInstance(this.id, this.duration, this.text, this.color, this.textStyle,
      this.tapAction, this.onCompletion);

  @override
  createState() => _BannerInstanceState();
}

class _BannerInstanceState extends State<_BannerInstance> {
  /// Whether the banner is being presented or removed
  bool isActive = true;

  /// The timer for removing banner from screen (if not tapped first)
  Timer timer;

  /// The height of the banner being presented (for animating it in/out)
  double bannerHeight;

  @override
  void initState() {
    super.initState();

    // Create the timer for dismissing banner after specified duration has passed
    timer = Timer(
      widget.duration,
      () {
        // Only dismiss if we have not already dismissed
        if (isActive) dismissAndDispose();
      },
    );

    // Get size of banner after first render pass
    WidgetsBinding.instance.addPostFrameCallback(
        (duration) => setState(() => this.bannerHeight = context.size.height));
  }

  void dismissAndDispose([TapUpDetails details]) {
    // Cancel delayed timer
    timer.cancel();

    // Dismiss banner
    setState(() => isActive = false);

    // Only call tapAction if it exists, and if banner was dismissed from a tap action
    if (widget.tapAction != null && details != null) widget.tapAction();

    // Remove as soon as animation is done
    Timer(
      Duration(milliseconds: 200),
      () => widget.onCompletion(widget.id),
    );
  }

  @override
  build(BuildContext context) {
    final top =
        bannerHeight == null ? -120.0 : (isActive ? 0.0 : -bannerHeight);

    return AnimatedPositioned(
      key: widget.key,
      left: 0,
      right: 0,
      top: top,
      duration: Duration(milliseconds: 180),
      child: GestureDetector(
        onTapUp: dismissAndDispose,
        onVerticalDragEnd: (details) {
          if (details.primaryVelocity < 0) dismissAndDispose();
        },
        child: Material(
          child: Container(
            padding: EdgeInsets.symmetric(horizontal: 16, vertical: 12),
            decoration: BoxDecoration(color: widget.color, boxShadow: [
              BoxShadow(
                color: Color(0x1F000000),
                offset: Offset(0, 1.75),
                blurRadius: 3.5,
              )
            ]),
            alignment: Alignment.center,
            child: SafeArea(
              bottom: false,
              child: Text(
                widget.text,
                style: TextStyle(
                        color: Colors.black,
                        fontSize: 15,
                        fontWeight: FontWeight.normal,
                        inherit: true)
                    .merge(widget.textStyle),
              ),
            ),
          ),
        ),
      ),
    );
  }
}
